lib/restrictedsession/audit.go

```diff
@@ -0,0 +1,155 @@
+// +build bpf,!386
+
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restrictedsession
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"unsafe"
+
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/bpf"
+	api "github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/trace"
+)
+
+const (
+	BlockedIP4 = iota
+	BlockedIP6
+)
+
+//
+// Audit Event types communicated between the kernel and userspace
+//
+
+// auditEventHeader matches audit_event_header in the C file
+type auditEventHeader struct {
+	CGroupID  uint64
+	PID       uint32
+	EventType int32
+	Command   [bpf.CommMax]byte
+}
+
+// auditEventBlockedIPv4 matches audit_event_blocked_ipv4 in the C file
+type auditEventBlockedIPv4 struct {
+	SrcIP   [4]byte
+	DstIP   [4]byte
+	DstPort uint16
+	Op      uint8
+}
+
+// auditEventBlockedIPv6 matches audit_event_blocked_ipv6 in the C file
+type auditEventBlockedIPv6 struct {
+	SrcIP   [16]byte
+	DstIP   [16]byte
+	DstPort uint16
+	Op      uint8
+}
+
+// newNetworkAuditEvent creates events.SessionNetwork, filling in common fields
+// from the SessionContext
+func newNetworkAuditEvent(ctx *bpf.SessionContext, hdr *auditEventHeader) events.SessionNetwork {
+	return events.SessionNetwork{
+		Metadata: events.Metadata{
+			Type: api.SessionNetworkEvent,
+			Code: api.SessionNetworkCode,
+		},
+		ServerMetadata: events.ServerMetadata{
+			ServerID:        ctx.ServerID,
+			ServerNamespace: ctx.Namespace,
+		},
+		SessionMetadata: events.SessionMetadata{
+			SessionID: ctx.SessionID,
+		},
+		UserMetadata: events.UserMetadata{
+			User:  ctx.User,
+			Login: ctx.Login,
+		},
+		BPFMetadata: events.BPFMetadata{
+			CgroupID: hdr.CGroupID,
+			Program:  bpf.ConvertString(unsafe.Pointer(&hdr.Command)),
+			PID:      uint64(hdr.PID),
+		},
+	}
+}
```

Comment:
@klizhentas You're reviewed the event structure?

Reply:
This reuses the existing event but added a `Restriction` field to indicate if it was allowed or denied. Enhanced Recording reports everything as `ALLOWED` and Restricted Session as `DENIED`. This causes an oddity in the UI. A denied action will end up with two audit log entries that look identical on the dashboard. One coming from the Enhanced Recording and one from Restricted Session. Furthermore, if you look at JSON details, the first one will show up as `Restriction=ALLOWED` and the second one as `Restriction=DENIED` :man_shrugging: